hc-nested-list-behavior.html

```diff
@@ -0,0 +1,178 @@
+<link rel="import" href="../polymer/polymer.html">
+
+<script>
+  /* global hc */
+  // Ensure our namespace exists
+  window.hc = window.hc || {};
+
+  /**
+   * Use `hc.NestedListBehavior` to implement a flat data structure of nested items into a list with
+   *  children getting added/remove to emulate open/close functionality.
+   *
+   * @polymerBehavior hc.NestedListBehavior
+   */
+  hc.NestedListBehavior = {
+    properties: {
+      /* The currently visible subset of items from the items property. */
+      _items: {
+        type: Array,
+        readOnly: true,
+        computed: '_computeItems(parts)'
+      },
+      /* Notifies if the element is in the process of loading, mainly used for lazy-loading. */
+      isLoading: {
+        type: Boolean,
+        value: false,
+        notify: true
+      },
+      /* The property name for the id on each item. */
+      itemIdName: {
+        type: String,
+        value: 'id'
+      },
+      /* The property name for the parentId on each item. */
+      itemParentIdName: {
+        type: String,
+        value: 'parentId'
+      },
+      /**
+       * The flat array of items to build the nested list structure from, requires an id and
+       *  parentId as defined by the itemIdName and itemParentIdName properties.
+       *
+       * @type {Object[]}
+       */
+      items: Array,
+      /**
+       * A nested representation of items, maintains list state; gets rebuilt when items changes,
+       *  when lazy-loading data add a new key-value pair to this object containing the data.  The
+       *  key value is used for looking up parentId to fetch the list of child items.  All item data
+       *  is taken from the items array.
+       */
+      parts: {
+        type: Object,
+        notify: true
+      }
+    },
+    observers: ['_computeParts(items)'],
+    listeners: {
+      toggle: 'toggleOpen'
+    },
+    /**
+     * Takes a parts object and optionally a start point for an array then iterates over the
+     *  object starting at the root if there isn't a location already set and finds all open nodes
+     *  recursively calling itself to ensure order is preserved once flattened.
+     *
+     * @param {Object} parts
+     * @param {Object[]} [items=parts.root]
+     * @param {String} [idName=this.itemIdName]
+     * @returns {Object[]}
+     */
+    _computeItems(parts, items = parts.root, idName = this.itemIdName) {
+      // Can't calculate items without parts or the name of the id property.
+      if (!idName || !parts) return [];
+
+      const newItems = [];
+
+      // Push all items into the new array, and if there is a child item then push all of the children.
+      items.forEach(item => {
+        item.open = Boolean(item.open);
```

Comment:
Yes, I tend to prefer the functionalities of the language over the side-effects of it's implementation when I can, in this case it also happens to make the code a lot clearer.